### PR TITLE
perf(profiling): datadog-alloc based StringTable

### PR DESCRIPTION
**Pull request by morrisonlevi**
# What does this PR do?

Creates a `StringTable` where the individual strings are allocated in
an arena allocator. This uses the Chain, Linear, and Virtual Allocators
provided by datadog-alloc to create a chain of 4 MiB chunks, each of
which is virtually allocated.

Also changes the chain allocator to handle large allocations, in case
we encounter a large string. Previously it would error if the
allocation size was larger than the node size.

# Motivation

This separates the StringTable from the system allocator. Since strings
are variable length, separating them from the system allocator has more
benefits than many other items.

This also is a bit faster. This is probably due to the simpler
allocation strategy and better data locality. See this for details:
https://github.com/DataDog/libdatadog/pull/404#discussion_r1600836784

# Additional Notes

This has been a long-time coming. I'm excited for it to finally merge!

# How to test the change?

As long as you weren't using string table implementation details,
nothing should change. The apis for FFI users is unchanged, for
instance.

### CODE REVIEW

File: profiling/benches/interning_strings.rs
Comment (ivoanjo) [inline profiling/benches/interning_strings.rs:26]: I'm curious: did you get a chance to compare this with the previous implementation?

File: profiling/src/collections/string_table/mod.rs
Comment (ivoanjo) [inline profiling/src/collections/string_table/mod.rs:38]: Minor: This is the kind of comment that tends to age poorly, e.g. maybe in some time we won't remember what this meant; may be worth expanding a bit ;)

File: profiling/src/collections/string_table/mod.rs
Comment (ivoanjo) [inline profiling/src/collections/string_table/mod.rs:43]: Minor:

```suggestion
        // It varies by implementation, but frequently I've noticed that the
```

File: profiling/src/collections/string_table/mod.rs
Comment (ivoanjo) [inline profiling/src/collections/string_table/mod.rs:57]: To be honest, I'd go even bigger -- 512 or maybe more? E.g. for one minute of data of a non-idle realistic application this seems like we'll hit it easily?

(I guess we could even try measuring this in staging or something)

File: profiling/src/collections/string_table/mod.rs
Comment (ivoanjo) [inline profiling/src/collections/string_table/mod.rs:109]: Minor: More of a structure kinda of thing, I'd kinda expect these two things (especially the unsafe bits) to be internal to `ChainAllocator`, rather than having the users need to get their hands dirty.

**Review by ivoanjo — COMMENTED** [review]
Oh wow, I have to say, after the PR with the really scary looking things for the allocator, this one is surprisingly clean and straightforward. (I always kinda fear a Rust PR will turn gnarly eventually lol)

I've left a few notes. I'd be curious to see the results of benchmarking this new implementation vs the previous one.

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs:57]: If only we had statistics on string data from libdatadog profiles...

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs:109]: In Rust, allocators don't initialize things. They deal with the shapes of the things (`Layout`) and give you bytes. The caller is responsible to initialize them.

I could create a custom `String` type that has a method like `String::new_in(str: &str, allocator: &impl Allocator)`. I don't feel like it's worth it for _this_ purpose only, but if there's some other reason you can think of to want it, then we could do that. We could bind the lifetime of the custom String struct, but then we'd have to deal with self-referential structs, which are also unsafe. The [ouroboros](https://crates.io/crates/ouroboros) crate can help with that, but it's a bit complex. We can do it if you or others think it's worth it.

File: profiling/benches/interning_strings.rs
Comment (morrisonlevi) [reply profiling/benches/interning_strings.rs:26]: I haven't yet, but intend to!

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs]: Thoughts on the reworded comment?

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs]: I did a little refactoring to pull the alloc + initialization into a single operation so the interning code reads in a better way. Let me know what you think. We can make these `'self -> 'static -> 'self` operations safe, if we really want to. It's just a lot of boilerplate to use ouroboros for it.

File: profiling/src/collections/string_table/mod.rs
Comment (ivoanjo) [reply profiling/src/collections/string_table/mod.rs:98]: I guess we could gather this if we wanted to -- we know in the backend which runtime a profile comes from, and so could even look at the data per-runtime.

File: profiling/src/collections/string_table/mod.rs
Comment (ivoanjo) [reply profiling/src/collections/string_table/mod.rs]: I think the new version is clearer, but I'm still kinda curious (and I don't think any of these are blockers, it's fine as-is):

* Currently the allocation exposes its guts. Then the caller uses them for what it wants. I was kinda wondering if it would make sense to do it the other way -- by telling the allocator: here's a pointer, and here's length, return me a pointer that's a result of copying the pointer/length inside yourself. Maybe even use generics to be able to return a pointer to whatever was passed in, rather than a generic pointer?

* The transmute in `intern` is it just a really fancy way of saying "never try to deallocate this string directly, as I'm managing the memory in another way"?

**Review by ivoanjo — COMMENTED** [review]
Just to clarify, is the idea not to merge this PR until we merge #402?

I think it's in pretty good state, but since a draft I'm not sure what plans you have and if I should approve as-is or wait for next steps, let me know! :pray:

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs]: - The allocation API is a mirror of what is in the work-in-progress Allocator API for standard Rust. It's not something I think we should change.
- The transmute of the lifetime doesn't deal with the deallocation in this case, because it's just a reference either way. The transmute is to tell the compiler that the lifetime of these references outlives the data it refers to. As mentioned, we can safely do this with self-referential lifetimes but Rust doesn't have those.

**Comment by morrisonlevi** [comment]
> Just to clarify, is the idea not to merge this PR until we merge [...]

Yes, although that is merged now and doesn't matter.

> I think it's in pretty good state, but since a draft I'm not sure what plans you have and if I should approve as-is or wait for next steps, let me know! 🙏

I think the code is overall in a good state. I've updated the PR description to match the state of current. There are still a few things like benchmarking the implementation vs the previous, as well as profiling the code to see if there is low hanging fruit. In particular, I wonder about `IndexMap` vs using a `Vec` + `HashMap`.

File: alloc/src/chain.rs
Comment (ivoanjo) [reply alloc/src/chain.rs:312]: Then we learn about it through our tests and update them accordingly? ;)

File: profiling/benches/interning_strings.rs
Comment (morrisonlevi) [reply profiling/benches/interning_strings.rs:26]: Alright, some numbers! Changing from the old implementation to the new one resulted in:

```
benching string interning on wordpress profile
                        time:   [53.695 µs 53.977 µs 54.279 µs]
                        change: [-28.080% -27.464% -26.850%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high severe
```

And then changing it back resulted in:

```
benching string interning on wordpress profile
                        time:   [75.224 µs 75.501 µs 75.820 µs]
                        change: [+40.269% +41.386% +42.503%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

This is unsurprising, because we spend less time in allocators, but I'm glad to have some data!

**Comment by morrisonlevi** [comment]
> In particular, I wonder about IndexMap vs using a Vec + HashMap.

It's worse than using `IndexMap`, which is good (you would hope a structure meant to solve this, does it well).

I added the old implementation for other people to A/B test if they want. Import whichever `StringTable` you want to use, and then do:

```
cargo bench -p datadog-profiling
```

Performance is improved. I used `perf` to look for low hanging fruit, there was nothing obvious to me there. We're ready for final review.

File: profiling/src/internal/profile.rs
Comment (danielsn) [inline profiling/src/internal/profile.rs:863]: Could this become a `.map().collect()`?

File: profiling/src/collections/string_table/wordpress_test_data.rs
Comment (danielsn) [inline profiling/src/collections/string_table/wordpress_test_data.rs:4]: A comment explaining where this came from, and how to modify it as needed, would be helpful to future maintainers.

File: profiling/benches/interning_strings.rs
Comment (danielsn) [inline profiling/benches/interning_strings.rs:45]: Make this a `cfg` rather than commenting out a line

File: profiling/src/collections/string_table/mod.rs
Comment (danielsn) [inline profiling/src/collections/string_table/mod.rs:277]: We're only adding each string once, in order.  Would it make sense to randomly add strings multiple times too?

File: profiling/src/collections/string_table/mod.rs
Comment (danielsn) [inline profiling/src/collections/string_table/mod.rs:17]: `allocate` is a funny name for this

File: profiling/src/collections/string_table/mod.rs
Comment (danielsn) [inline profiling/src/collections/string_table/mod.rs:62]: Should we take a lifetime paramater / is there way to tie this to the lifetime of the underlying object?

File: profiling/src/collections/string_table/mod.rs
Comment (danielsn) [inline profiling/src/collections/string_table/mod.rs:51]: Where do we enforce that the first element is always `""`

File: profiling/src/collections/string_table/mod.rs
Comment (danielsn) [reply profiling/src/collections/string_table/mod.rs:98]: The cost of reserving a bit too much shouldn't be that bad

File: profiling/src/collections/string_table/mod.rs
Comment (danielsn) [inline profiling/src/collections/string_table/mod.rs:102]: Seems like a worthwhile special case to handle in the allocator.

File: profiling/src/collections/string_table/mod.rs
Comment (danielsn) [inline profiling/src/collections/string_table/mod.rs:110]: Since `is_empty` would be `false`?

File: profiling/src/collections/string_table/mod.rs
Comment (danielsn) [inline profiling/src/collections/string_table/mod.rs:120]: I know errors are annoying to propagate, but panic will bring down the customer process.

File: profiling/src/collections/string_table/mod.rs
Comment (danielsn) [inline profiling/src/collections/string_table/mod.rs:145]: is it possible to give it the same lifetime as `self`?

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs:120]: Yes, I am aware, but this is just like the `GlobalAlloc` panic'ing if it runs out of memory. Same fringe error condition.

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs:145]: There is no such thing as a `'self` lifetime, unfortunately 🙁

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs:110]: Yes, it doesn't make sense to ask the question "is the string table empty?" because it's never empty.

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs:102]: The trouble is, what allocators are supposed to do for zero-sized allocations is not pinned down. In C, you get implementation defined behavior for `malloc(0)` and it's generally recommended to avoid it because of this. The debate is still open for Rust.

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs:17]: Have any ideas for a better one? If I had a better one, I'd have used it ^_^

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs:62]: Yes, there is a way to do it using `ouroboros`. The code to do it is quite ugly, because they have to restrict all kinds of things to make it safe. Under the hood, it's basically doing the same trick.

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs:51]: It's in `StringTable::new`, a few lines down.

File: profiling/src/collections/string_table/mod.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/mod.rs:277]: The `test_from_src` function adds every string twice.

File: profiling/src/internal/profile.rs
Comment (morrisonlevi) [reply profiling/src/internal/profile.rs:863]: The lending iterator doesn't have it implemented. I could do that if you want, but I'd prefer to do it in another PR. Let me know.

File: profiling/src/collections/string_table/wordpress_test_data.rs
Comment (morrisonlevi) [reply profiling/src/collections/string_table/wordpress_test_data.rs:4]: 👍🏻 I'll move the comment from `test_wordpress` to the top of this file.

File: profiling/benches/interning_strings.rs
Comment (morrisonlevi) [reply profiling/benches/interning_strings.rs:45]: When you are benchmarking you generally make tweaks to things and run it again. It's not just a static A/B thing we're going to run in CI so you want to automate swapping it out.. So it seems weird to have a cfg for this reason.

I think to use the cfg without making code changes, that would mean making a feature for it. Having a feature just for a benchmarking thing seems odd. Plus it breaks the idea of feature unification (features should be additive).

Do these things make sense? Thoughts?

**Review by r1viollet — APPROVED** [review]
LGTM
We should be careful about performance measures while deploying this.
The CI failure is not relevant to this PR.

**Review by danielsn — APPROVED** [review]
LGTM

File: profiling/src/internal/profile.rs
Comment (morrisonlevi) [reply profiling/src/internal/profile.rs:863]: Doing `map` is possible it has usability issues. `collect` seems even harder, because you need to have a `Map` or something which breaks the borrow of the item (e.g. `String::from`) because otherwise the borrow prevents the lending iterator from moving to the next item (because it's `mut`).

**Pull request merged**
